python/shotgun_search_widget/shotgun_search_widget.py

```diff
@@ -0,0 +1,126 @@
+# Copyright (c) 2015 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+from sgtk.platform.qt import QtCore, QtGui
+
+
+class ShotgunSearchWidget(QtGui.QLineEdit):
+    """
+    A QT Widget deriving from :class:`~PySide.QtGui.QLineEdit` that creates
+    a search input box with auto completion.
+
+    The derived classes are expected to provide a :class:`PySide.QtGui.QCompleter`
+    during initialization. The completer must have ``search(str)`` and ``destroy`` method.
+    """
+    def __init__(self, parent):
+        """
+        :param parent: Qt parent object
+        :type parent: :class:`~PySide.QtGui.QWidget`
+        """
+
+        # first, call the base class and let it do its thing.
+        super(ShotgunSearchWidget, self).__init__(parent)
+
+        # trigger the completer to popup as text changes
+        self.textEdited.connect(self._search_edited)
+
+        # Taken from https://wiki.qt.io/Delay_action_to_wait_for_user_interaction
+        self._delay_timer = QtCore.QTimer(self)
+        self._delay_timer.timeout.connect(self._typing_timeout)
+        self._delay_timer.setSingleShot(True)
+
+        # FIXME: The following was stolen from SearchWidget. We can't refactor easily that
+        # part of the code since the base classes for ShotgunSearchWidget and SearchWidget
+        # are not the same, but at least the ShotgunSearchWidget has feature parity.
+        self.set_placeholder_text("Search")
+
+        # dynamically create the clear button so that we can place it over the
+        # edit widget:
+        self._clear_btn = QtGui.QPushButton(self)
+        self._clear_btn.setFocusPolicy(QtCore.Qt.StrongFocus)
+        self._clear_btn.setFlat(True)
+        self._clear_btn.setCursor(QtCore.Qt.ArrowCursor)
+        style = ("QPushButton {"
```

Comment:
As the comment says above, I copy pasted this bit from somewhere else. I'll fix it.